### Segmenting and Clustering Neighborhoods in Toronto
In this notebook, I am going to scrap data from wikipedia to get the list of neighborhoods in Toronto and then use Square API to cluster these neighborhoods based on the venues they contain.

## Part 1 : Getting the neighborhoods of Toronto

In [1]:
import pandas as pd

In [2]:
from IPython.display import IFrame
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
IFrame(url, width=800, height=350)

#### Using this wikipedia webpage, we will create a dataframe containing three columns: PostalCode, Borough, and Neighborhood

In [3]:
data, = pd.read_html(url, match="Postal", skiprows=1)
data.columns = ["PostalCode", "Borough", "Neighborhood"]
data.head()

PostalCode           Borough                      Neighborhood
0        M2A      Not assigned                               NaN
1        M3A        North York                         Parkwoods
2        M4A        North York                  Victoria Village
3        M5A  Downtown Toronto         Regent Park, Harbourfront
4        M6A        North York  Lawrence Manor, Lawrence Heights

Let's drop all the cells that don't have any Borough assigned to them:

In [4]:
data = data[data["Borough"] != "Not assigned"]
data.head()

PostalCode           Borough                                 Neighborhood
1        M3A        North York                                    Parkwoods
2        M4A        North York                             Victoria Village
3        M5A  Downtown Toronto                    Regent Park, Harbourfront
4        M6A        North York             Lawrence Manor, Lawrence Heights
5        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [5]:
data['PostalCode'].describe()

count     103
unique    103
top       M1R
freq        1
Name: PostalCode, dtype: object

As we can notice, there is 103 different Postal Code. 
Each Postal code is associated with a Borough and a certain number of neighborhoods seperated by a comma.

Fortunately, each Borough has at least a neighborhood associated with it, so we don't need further data cleaning.

In [6]:
data[data["Neighborhood"] == "Not assigned"]

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [7]:
data[data["Neighborhood"] == ""]

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [8]:
data.shape

(103, 3)

In [9]:
#import sys
#!conda install -c conda-forge geopy --yes --prefix {sys.prefix}
#!conda install -c conda-forge folium=0.5.0 --yes --prefix {sys.prefix}
#!conda install -c conda-forge geocoder --yes --prefix {sys.prefix}

We will import the coordinates associated with each postal code in Toronto using the csv table in the following link:

In [10]:
locations = pd.read_csv("https://cocl.us/Geospatial_data")
locations.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [11]:
locations.columns = ["PostalCode", "Latitude", "Longitude"]
data = pd.merge(data, locations, on='PostalCode')
data.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [12]:
data[data["PostalCode"] == "M6A"]

PostalCode     Borough                      Neighborhood   Latitude  \
3        M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   

   Longitude  
3 -79.464763

## Part 2 : Venues of Toronto

In [16]:
import pandas as pd
import numpy as np
import json

#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

import requests
from pandas.io.json import json_normalize

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes
import folium

In [32]:
Toronto  = data[data['Borough'].str.contains("Toronto")]
Toronto .head()

PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031

In [33]:
from geopy.geocoders import Nominatim
address = 'Toronto'
geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {},{}.'.format(latitude,longitude))

The geograpical coordinate of Toronto are 43.6534817,-79.3839347.


/home/hassen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.22.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


## Toronto Map

In [35]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Toronto['Latitude'], Toronto['Longitude'], Toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

## Foursquare credentials

In [36]:
CLIENT_ID = 'aa' # your Foursquare ID
CLIENT_SECRET = 'aa' # your Foursquare Secret
VERSION = '20200202' # Foursquare API version

## Exploring other neighborhood

In [37]:
Etobicoke = data[data['Borough']=='Etobicoke'].reset_index(drop=True)

# get the Etobicoke latitude and longitude values
Etobicoke_latitude = Etobicoke.loc[0,'Latitude']
Etobicoke_longitude = Etobicoke.loc[0,'Longitude']

print('Latitude and longitude values of Etobicoke are {},{}.'.format(
    Etobicoke_latitude, Etobicoke_longitude))

Latitude and longitude values of Etobicoke are 43.6678556,-79.53224240000002.


In [38]:
map_Etobicoke = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Etobicoke['Latitude'], Etobicoke['Longitude'], Etobicoke['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Etobicoke)  
    
map_Etobicoke

## Getting the top 100 venues that are in Toronto within a raidus of 500 meters

In [40]:
LIMIT = 100
radius = 500

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


# run the above function
Toronto_venues = getNearbyVenues(names=Toronto['Neighborhood'],
                                  latitudes=Toronto['Latitude'],
                                  longitudes=Toronto['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town,

## Resulting dataframe

In [42]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1596, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

## Check how many venues were returned

In [43]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre                                  18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          22   
India Bazaar, The Beaches West                                         24   
Kensington Market, Chinatown, Grange Park                              55   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     21   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              47   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     37   
St. James Town                                                         76   
St. James Town, Cabbagetown                                            41   
Stn A PO Boxes                                                         94   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre                                   18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

## Find out how many unique categories can be curated from all the returned venue

In [44]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 234 uniques categories.


## Analyze Neighborhood : One-hot encoding

In [47]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

print(Toronto_onehot.shape)
Toronto_onehot.head()

(1596, 234)


Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Toy / Game Store  Trail  Train Station  \
0             0  ...                 0      0              0   
1             0  ...                 0      0              0   
2             0  ...                 0      0              0   
3             0  ...                 0      0              0   
4             0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  
0                      0         0            0              0  
1                      0         0            0              0  
2                      0         0            0              0  
3                      0         0            0              0  
4                      0         0            0              0  

[5 rows x 234 columns]

## Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [50]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
print(Toronto_grouped.shape)
Toronto_grouped.head()

(39, 234)


Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1       Brockton, Parkdale Village, Exhibition Place     0.000000   
2              Business reply mail Processing Centre     0.055556   
3  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                 Central Bay Street     0.015873   

   Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0                0.0  0.000000            0.000000      0.000000   
1                0.0  0.000000            0.000000      0.000000   
2                0.0  0.000000            0.000000      0.000000   
3                0.0  0.058824            0.058824      0.058824   
4                0.0  0.000000            0.000000      0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                  0.0   
1        0.000000         0.000000          0.000000                  0.0   
2        0.000000         0.000000          0.000000                  0.0   
3        0.117647         0.117647          0.117647                  0.0   
4        0.000000         0.000000          0.000000                  0.0   

   ...  Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0  ...               0.0    0.0            0.0                       0.018182   
1  ...               0.0    0.0            0.0                       0.000000   
2  ...               0.0    0.0            0.0                       0.000000   
3  ...               0.0    0.0            0.0                       0.000000   
4  ...               0.0    0.0            0.0                       0.015873   

   Video Game Store  Video Store  Vietnamese Restaurant  Wine Bar  \
0               0.0     0.000000                    0.0       0.0   
1               0.0     0.000000                    0.0       0.0   
2               0.0     0.000000                    0.0       0.0   
3               0.0     0.000000                    0.0       0.0   
4               0.0     0.015873                    0.0       0.0   

   Wings Joint  Women's Store  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0  

[5 rows x 234 columns]

## Pick up top 10 venues

In [53]:
num_top_venues = 10

# write a funcition to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)   
    return row_categories_sorted.index.values[0:num_top_venues]

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Toronto_venues_sorted = pd.DataFrame(columns=columns)
Toronto_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    Toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

Toronto_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing Centre    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                Bakery   
1        Breakfast Spot           Coffee Shop         Grocery Store   
2           Yoga Studio         Auto Workshop                  Park   
3       Airport Service      Airport Terminal           Coffee Shop   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop                  Café   
1                Bakery               Stadium         Burrito Place   
2            Comic Shop           Pizza Place      Recording Studio   
3         Boat or Ferry   Rental Car Location      Sculpture Garden   
4                   Bar       Bubble Tea Shop        Ice Cream Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Grocery Store    Japanese Restaurant  
1            Restaurant          Climbing Gym              Pet Store  
2         Burrito Place            Restaurant                Brewery  
3                 Plane                   Bar        Harbor / Marina  
4           Salad Place   Japanese Restaurant        Thai Restaurant

# Part3: Cluster Neighborhoods

In [82]:
# set number of clusters
kclusters = 3

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [83]:
Toronto_merged = Toronto

# add clustering labels
Toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(Toronto_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head()

/home/hassen/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               0           Coffee Shop   
4   43.662301 -79.389494               0           Coffee Shop   
9   43.657162 -79.378937               0        Clothing Store   
15  43.651494 -79.375418               0           Coffee Shop   
19  43.676357 -79.293031               0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                Bakery                   Pub   
4       Sushi Restaurant           Yoga Studio              Creperie   
9            Coffee Shop                  Café       Bubble Tea Shop   
15                  Café          Cocktail Bar   American Restaurant   
19                   Pub     Health Food Store               Dog Run   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
2         Breakfast Spot                  Café                    Theater   
4               Beer Bar         Smoothie Shop             Sandwich Place   
9    Japanese Restaurant    Italian Restaurant  Middle Eastern Restaurant   
15             Gastropub              Beer Bar         Seafood Restaurant   
19      Department Store          Dessert Shop                      Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2         Cosmetics Shop            Shoe Store             Restaurant  
4          Burrito Place                  Café     College Auditorium  
9         Cosmetics Shop            Restaurant                  Diner  
15              Creperie                   Gym         Lingerie Store  
19        Discount Store   Distribution Center          Women's Store

In [84]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], 
                                  Toronto_merged['Neighborhood'], 
                                  Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 1:
#### This seems to be a cluster of mix retail shops

In [85]:
# Cluster 1 
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto               0           Coffee Shop   
4    Downtown Toronto               0           Coffee Shop   
9    Downtown Toronto               0        Clothing Store   
15   Downtown Toronto               0           Coffee Shop   
19       East Toronto               0                 Trail   
20   Downtown Toronto               0           Coffee Shop   
24   Downtown Toronto               0           Coffee Shop   
25   Downtown Toronto               0         Grocery Store   
30   Downtown Toronto               0           Coffee Shop   
31       West Toronto               0                Bakery   
36   Downtown Toronto               0           Coffee Shop   
37       West Toronto               0                   Bar   
41       East Toronto               0      Greek Restaurant   
42   Downtown Toronto               0           Coffee Shop   
43       West Toronto               0                  Café   
47       East Toronto               0                  Park   
48   Downtown Toronto               0           Coffee Shop   
54       East Toronto               0                  Café   
62    Central Toronto               0                Garden   
68    Central Toronto               0                  Park   
69       West Toronto               0       Thai Restaurant   
73    Central Toronto               0        Clothing Store   
74    Central Toronto               0                  Café   
75       West Toronto               0             Gift Shop   
81       West Toronto               0           Coffee Shop   
83    Central Toronto               0                  Park   
84   Downtown Toronto               0                  Café   
86    Central Toronto               0                   Pub   
87   Downtown Toronto               0        Airport Lounge   
91   Downtown Toronto               0                  Park   
92   Downtown Toronto               0           Coffee Shop   
96   Downtown Toronto               0           Coffee Shop   
97   Downtown Toronto               0           Coffee Shop   
99   Downtown Toronto               0   Japanese Restaurant   
100      East Toronto               0    Light Rail Station   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                    Park                Bakery                   Pub   
4        Sushi Restaurant           Yoga Studio              Creperie   
9             Coffee Shop                  Café       Bubble Tea Shop   
15                   Café          Cocktail Bar   American Restaurant   
19                    Pub     Health Food Store               Dog Run   
20           Cocktail Bar              Beer Bar                Bakery   
24     Italian Restaurant        Sandwich Place                  Café   
25                   Café                  Park           Candy Store   
30                   Café            Restaurant         Deli / Bodega   
31               Pharmacy         Grocery Store                   Bar   
36               Aquarium                  Café                 Hotel   
37             Restaurant      Asian Restaurant           Men's Store   
41     Italian Restaurant           Coffee Shop            Restaurant   
42                   Café                 Hotel            Restaurant   
43         Breakfast Spot           Coffee Shop         Grocery Store   
47   Fast Food Restaurant        Sandwich Place           Coffee Shop   
48                   Café            Restaurant                 Hotel   
54            Coffee Shop                Bakery             Gastropub   
62          Women's Store          Dance Studio  Ethiopian Restaurant   
68          Jewelry Store                 Trail      Sushi Restaurant   
69     Mexican Restaurant                  Café                  Park   
73    Sporting Goods Shop           Coffee Shop  Gym / Fitness Center   
74         Sandwich Place           Coffee Shop         Grocery St

## Cluster 2:
#### This seems to be a cluster of Parks and non eating shops

In [86]:
# Cluster 2
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61  Central Toronto               1                  Park   
67  Central Toronto               1                  Park   
79  Central Toronto               1        Sandwich Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61           Swim School              Bus Line  Colombian Restaurant   
67        Breakfast Spot      Department Store                 Hotel   
79          Dessert Shop                  Café           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
61      Department Store  Ethiopian Restaurant     Electronics Store   
67     Food & Drink Shop        Sandwich Place          Dance Studio   
79    Italian Restaurant      Sushi Restaurant                   Gym   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
61  Eastern European Restaurant            Donut Shop       Doner Restaurant  
67                          Gym               Dog Run          Deli / Bodega  
79                  Pizza Place                 Diner                    Bar

## Cluster 3:
#### This seems to be a cluster of Cafés and bars

In [87]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] ==2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
80  Downtown Toronto               2                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
80   Japanese Restaurant                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
80             Bookstore            Restaurant                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
80    Chinese Restaurant                   Pub           Dessert Shop